File: character.cpp
#include "character.h"
#include "gun.h"
#include "lasergun.h"
#include "grenadegun.h"

qint32 sign(const qint32 &x) {
    return (x > 0) - (x < 0);
}

Character::Character(Map *map, Observer *view, QQuickItem *item, QPoint position) : m_map(map), m_view(view) {
    m_item = item;
    m_health = 100;
    m_mana = 100;
    m_goingLeft = m_goingRight = false;
    m_jumping = false;
    m_vspeed = 0;
    m_coord = position;
    m_reload = 0;
    m_weapon = new Gun();
    m_weaponType = 0;
    m_boundingBox = QRect(m_coord.x(), m_coord.y(), 55, 85);
    m_spriteFlipped = false;
}

Character::~Character() {
    delete m_weapon;
}

void Character::goLeft() {
    m_goingLeft = true;
}

void Character::stopLeft() {
    m_goingLeft = false;
}

void Character::goRight() {
    m_goingRight = true;
}

void Character::stopRight() {
    m_goingRight = false;
}

void Character::jump() {
    if (m_map->isFilled(m_boundingBox.translated(0, 1)))
        m_vspeed = 30;
}

void Character::changeWeapon() {
    delete m_weapon;
    m_weaponType = (m_weaponType + 1) % 3;
    switch (m_weaponType) {
        case 0: {
            m_weapon = new Gun();
            break;
        }
        case 1: {
            m_weapon = new LaserGun();
            break;
        }
        case 2: {
            m_weapon = new GrenadeGun();
            break;
        }
    }
}

void Character::attack(qint32 mouseX, qint32 mouseY) {
    if (m_reload == 0 and m_mana >= m_weapon->getManaCost()) {
        m_reload = 20;
        setMana(m_mana - m_weapon->getManaCost());
        QVector2D startCoord = getHandPosition();
        m_weapon->shoot(m_view, QVector2D(mouseX, mouseY), startCoord, m_map, m_id);
    }
}

void Character::hit(qreal value) {
    m_health = std::max(0.0, m_health - value);
}

bool Character::update() {
    m_health = std::min(m_health + 0.1, 100.0);
    setMana(std::min(m_mana + 1, 100));
    m_vspeed = std::max(m_vspeed - 1, 0);
    m_reload = std::max(m_reload - 1, 0);

    if (m_goingLeft xor m_goingRight) {
        if (m_goingLeft) {
            if (not m_spriteFlipped)
                flipSprite();
            moveHorizontal(-5);
        }
        if (m_goingRight) {
            if (m_spriteFlipped)
                flipSprite();
            moveHorizontal(5);
        }
    }
    m_inAir = (not m_map->isFilled(m_boundingBox.translated(0, 1)));
    moveVertical(10 - m_vspeed);

    m_item->setPosition(m_coord);
    m_item->setProperty("mirrored", m_spriteFlipped);
    m_item->setProperty("going", m_goingLeft xor m_goingRight);
    m_item->setProperty("inAir", m_inAir);
    m_item->setProperty("health", m_health);
    m_item->setProperty("mana", m_mana);
    m_item->setProperty("attack", m_reload != 0);
    return false;
}

void Character::moveHorizontal(qint32 speed) {
    qint32 dx = speed;
    qint32 delta = sign(speed);
    QRect newRect = m_boundingBox.translated(dx, 0);
    while (m_map->isFilled(newRect) and dx != 0) {
        dx -= delta;
        newRect.translate(-delta, 0);
    }
    if (dx == 0)
        return;
    m_coord.setX(m_coord.x() + dx);
    m_boundingBox.translate(dx, 0);
}

void Character::moveVertical(qint32 speed) {
    qint32 dy = speed;
    qint32 delta = sign(speed);
    QRect newRect = m_boundingBox.translated(0, dy);
    while (m_map->isFilled(newRect) and dy != 0) {
        dy -= delta;
        newRect.translate(0, -delta);
        m_vspeed = 0;
    }
    if (dy == 0)
        return;
    m_inAir = true;
    m_coord.setY(m_coord.y() + dy);
    m_boundingBox.translate(0, dy);
}

QVector2D Character::getHandPosition() {
    const int offsetX = 8;
    const int offsetY = 25;

    if (m_spriteFlipped)
        return QVector2D(m_coord.x() + offsetX, m_coord.y() + offsetY);
    else
        return QVector2D(m_coord.x() + 55 - offsetX, m_coord.y() + offsetY);
}

File: src/character.ts
import {Vec2, Rect, sign, rect, rectTranslated, clamp, ID, add, newEntity} from "./types.ts";
import {GameMap} from "./map.ts";
import {WeaponType, mana_cost, reload} from "./weapon.ts";
import {World} from "./world.ts";

export class Character {
  protected m_id: ID;
  m_coord: Vec2 = { x: 0, y: 0 };
  protected m_vspeed = 0;
  health = 100;
  mana = 100;
  score = 0;
  m_reload = 0;
  boundingBox: Rect = { x: 0, y: 0, w: 55, h: 90 };
  weaponId: WeaponType;
  m_goingLeft = false;
  m_goingRight = false;
  mirrored = false;
  inAir = true;

  constructor(position: Vec2, id: ID) {
    this.m_coord = {...position};
    this.boundingBox = rect(position, 55, 90);
    this.weaponId = WeaponType.Bullet;
    this.m_id = id;
  }

  // alias for m_coord so world/renderer can read position
  get x(): number { return this.m_coord.x; }
  get y(): number { return this.m_coord.y; }

  goLeft(): void { this.m_goingLeft = true; }
  stopLeft(): void { this.m_goingLeft = false; }
  goRight(): void { this.m_goingRight = true; }
  stopRight(): void { this.m_goingRight = false; }

  jump(map: GameMap): void {
    if (map.isFilledRect(rectTranslated(this.boundingBox, 0, 1)))
      this.m_vspeed = 30;
  }

  changeWeapon(): void {
    this.weaponId = (this.weaponId + 1) % 3 as WeaponType;
  }

  attack(w: World, p: Vec2): void {
    const mana = mana_cost[this.weaponId];
    if (this.m_reload === 0 && this.mana >= mana) {
      this.m_reload = reload[this.weaponId];
      this.mana -= mana;
      const hand = this.getHandPosition();
      w.component_weapon.set(newEntity(), {
        type: this.weaponId,
        mana: mana,
        owner_id: this.m_id,
        pos: hand,
        target: p,
      });
    }
  }

  hit(value: number): void {
    this.health = Math.max(0, this.health - value);
  }

  respawn(p: Vec2): void {
    this.m_coord = {...p};
    this.boundingBox = rect(p, 55, 90);
    this.health = 100;
    this.mana = 100;
    this.m_goingLeft = this.m_goingRight = false;
    this.m_vspeed = 0;
    this.m_reload = 0;
    this.mirrored = false;
  }

  get id(): ID { return this.m_id; }
  addKill(): void { this.score++; }

  update(map: GameMap): void {
    this.health = clamp(this.health + 0.2, 0, 100);
    this.mana = clamp(this.mana + 0.8, 0, 100);
    this.m_vspeed = Math.max(this.m_vspeed - 1, 0);
    this.m_reload = Math.max(this.m_reload - 1, 0);

    if (this.m_goingLeft !== this.m_goingRight) {
      if (this.m_goingLeft) {
        if (!this.mirrored)
          this.flipSprite();
        this.moveHorizontal(map, -5);
      }
      if (this.m_goingRight) {
        if (this.mirrored)
          this.flipSprite();
        this.moveHorizontal(map, 5);
      }
    }
    this.inAir = !map.isFilledRect(rectTranslated(this.boundingBox, 0, 1));
    this.moveVertical(map, 10 - this.m_vspeed);
  }

  protected flipSprite(): void {
    this.mirrored = !this.mirrored;
  }

  protected moveHorizontal(map: GameMap, speed: number): void {
    let dx = speed;
    const delta = sign(speed);
    let newRect = rectTranslated(this.boundingBox, dx, 0);
    while (map.isFilledRect(newRect) && dx !== 0) {
      dx -= delta;
      newRect = rectTranslated(newRect, -delta, 0);
    }
    if (dx === 0)
      return;
    this.m_coord.x += dx;
    this.boundingBox.x += dx;
  }

  protected moveVertical(map: GameMap, speed: number): void {
    let dy = speed;
    const delta = sign(speed);
    let newRect = rectTranslated(this.boundingBox, 0, dy);
    while (map.isFilledRect(newRect) && dy !== 0) {
      dy -= delta;
      newRect = rectTranslated(newRect, 0, -delta);
      this.m_vspeed = 0;
    }
    if (dy === 0)
      return;
    this.inAir = true;
    this.m_coord.y += dy;
    this.boundingBox.y += dy;
  }

  getHandPosition(): Vec2 {
    const o = {x: 8, y: 25};
    return add(this.m_coord, {x: this.mirrored ? o.x : 55 - o.x, y: o.y});
  }

  get going(): boolean {
    return this.m_goingLeft !== this.m_goingRight;
  }

  get attacking(): boolean {
    return this.m_reload !== 0;
  }
}
